### PR TITLE
Add Autocast support for Einsum

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
ONNX spec for Einsum requires all inputs to be the same dtype.

PyTorch runtime does not allow executing aten::einsum with
mismatching types by default, so the export would never succeed,

However, when the model is wrapped by `torch.autocast()`,
the run succeeds and the ONNX converter will create an Einsum ONNX node 
with mismatch types as input, which is not allowed by the aforementioned schema.

This PR adds onnx::Einsum to the Autocast enabled list, which outputs lower resolution tensors

### CODE REVIEW

File: test/onnx/test_pytorch_onnx_onnxruntime.py
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:5146]: The exported ONNX model output is fp32, while pytorch output is bfloat16. I'd recommend updating `ort_compare_with_pytorch` to also check if dtype matches.

Do we know how `torch.autocast` works underneath? It does not insert cast into the traced graph, but somehow end up getting inputs casted.

File: test/onnx/test_pytorch_onnx_onnxruntime.py
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:67]: Any idea why inputs of type float16 and float creates output of type bfloat16?

Comment (thiagocrepaldi) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:67]: Probably because bfloat16 has a higher range than float16 and it is m=faster to convert float32 to bfloat16

https://nhigham.com/2018/12/03/half-precision-arithmetic-fp16-versus-bfloat16/ has a nice write up

File: test/onnx/test_pytorch_onnx_onnxruntime.py
Comment (thiagocrepaldi) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:5146]: I didnt look into much detail, but autocast is a kind of context manager to store the actual dtype used during computation while preserving the original inputs dtype in the aten graph

**Comment by BowenBao** [comment]
A few concerns over the type issue here.

1. The exported model does not match in datatype. PyTorch autocast runs Einsum with inputs in fp16, outputing bfloat16. ONNX model runs Einsum(Matmul) with inputs in fp32, outputing fp32.
2. `scalar_type_analysis` pass was designed to cover standard PyTorch implicit cast. This use case is somewhat out of scope. Or at least the pass needs some extra extension, the reason being:
    * The casting here is introduced by autocast. Without autocast, PyTorch itself will fail when `Einsum` or `Matmul` receives inputs of mixed datatype. 
    * The casting logic is different vs standard pytorch implicit cast. For operator receiving inputs of both fp32 and fp16 types, [Standard](https://pytorch.org/docs/stable/tensor_attributes.html#id2) casts fp16 to fp32, [autocast ](https://pytorch.org/docs/stable/amp.html#op-specific-behavior) casts fp32 to fp16, based on operator type. I feel it is better to not mixed the two.

If this short term solution is essential and urgent for customer unblocking, please add comments noting the above points in the code. This comment applies similarly to the batchnorm PR #72565

cc @garymm

**Comment by garymm** [comment]
Agreed with Bowen: I think we need to:
* understand what's going on here at a deeper level. Why does PT output bfloat16 in this case?
* try to make the converted model match the PyTorch behavior whenever possible.

AFAIK this isn't blocking a production launch so I think we can take the time to do this investigation, but if there is a need for urgency we can move forward with this PR with lots of TODOs inserted.

**Comment by thiagocrepaldi** [comment]
By default, `torch.autocast` use `bloat16` when device is cpu and float16 otherwise
```
// autocast_cpu_dtype is the lower_precision_fp used by AutocastCPU.
thread_local at::ScalarType autocast_cpu_dtype = at::kBFloat16;

// autocast_gpu_dtype is the lower_precision_fp used by AutocastGPU.
thread_local at::ScalarType autocast_gpu_dtype = at::kHalf;
```

**Comment by thiagocrepaldi** [comment]
> Agreed with Bowen: I think we need to:
> 
> * understand what's going on here at a deeper level. Why does PT output bfloat16 in this case?
> * try to make the converted model match the PyTorch behavior whenever possible.
> 
> AFAIK this isn't blocking a production launch so I think we can take the time to do this investigation, but if there is a need for urgency we can move forward with this PR with lots of TODOs inserted.

I will keep this reply updated with my findings:

By default, `torch.autocast` uses `bloat16` when the tensor device is CPU and float16 otherwise
```
// autocast_cpu_dtype is the lower_precision_fp used by AutocastCPU.
thread_local at::ScalarType autocast_cpu_dtype = at::kBFloat16;

// autocast_gpu_dtype is the lower_precision_fp used by AutocastGPU.
thread_local at::ScalarType autocast_gpu_dtype = at::kHalf;
```
As the proposed tests were for CPU, the output was downcasted to bfloat16. There are mechanisms to force a different type, but it wasnt used in the repro.

**Comment by garymm** [comment]
If this isn't ready for review, please convert to draft.

**Comment by garymm** [comment]
I'll let @BowenBao continue the review, since he is the assignee. Bowen please discuss in our next stand-up if you have concerns still.

File: aten/src/ATen/core/interned_strings.h
Comment (BowenBao) [inline aten/src/ATen/core/interned_strings.h:223]: I suppose this is no longer needed?

File: aten/src/ATen/autocast_mode.cpp
Comment (BowenBao) [inline aten/src/ATen/autocast_mode.cpp:328]: We'd probably need review from Meta folks on this part. 

The first question on top of my head is that is it legit to apply the same casting rule for all different einsum formulas?

**Comment by BowenBao** [comment]
@thiagocrepaldi I think we should still add the unittest in `test_pytorch_onnx_onnxruntime.py`, and update the result comparison function to check for dtype as well. (bfloat16 could be relaxed when comparing to fp16 for now as we improve)

@msaroufim Do you know who works on autocast that could take a look at this PR?

File: aten/src/ATen/core/interned_strings.h
Comment (thiagocrepaldi) [reply aten/src/ATen/core/interned_strings.h:223]: I asked myself the same question. What is the policy for adding interned strings? Should every onnx op be present here as they are added to onnx spec or just when they are being referenced by code?

File: aten/src/ATen/autocast_mode.cpp
Comment (thiagocrepaldi) [reply aten/src/ATen/autocast_mode.cpp:328]: Einsum fundamentally does multiplication and addition. The equations merely say which dimensions are multiplied together. The dimensions that are not multiplied are then added together

As both tensor multiplication and summation are both present in the autocast, it seems reasonable that an op that does both operations in a single go also be present.

I am OK in double checking with Meta, though.

**Comment by msaroufim** [comment]
> @msaroufim Do you know who works on autocast that could take a look at this PR?

@gottbrath @jerryzh168 would one of you happen to know who is the right person to review this PR?

File: aten/src/ATen/autocast_mode.cpp
Comment (thiagocrepaldi) [reply aten/src/ATen/autocast_mode.cpp:328]: Added `onnx-needs-import` to get on Meta radar. Is that ok?

**Review by BowenBao — APPROVED** [review]

**Comment by garymm** [comment]
@thiagocrepaldi is the PR description still accurate? It says:

> This PR adds onnx::Einsum to the Scalar Type Analysis pass that promotes different input types so that they match

I thought the current approach doesn't involve ONNX at all?

**Comment by thiagocrepaldi** [comment]
> @thiagocrepaldi Thiago Crepaldi FTE is the PR description still accurate? It says:
> 
> > This PR adds onnx::Einsum to the Scalar Type Analysis pass that promotes different input types so that they match
> 
> I thought the current approach doesn't involve ONNX at all?

it doesnt. outdated pr description. will fix it

**Comment by thiagocrepaldi** [comment]
> > @msaroufim Do you know who works on autocast that could take a look at this PR?
> 
> @gottbrath @jerryzh168 would one of you happen to know who is the right person to review this PR?

Gentle ping @msaroufim @gottbrath @jerryzh168

**Comment by malfet** [comment]
@ngimel do you mind taking a look at this one?

**Review by ngimel — APPROVED** [review]

**Comment by ngimel** [comment]
lgtm

**Comment by thiagocrepaldi** [comment]
@malfet This PR was approved by @ngimel 
Do you mind importing it?

**Pull request closed without merging**
